prosemirror/model/from_dom.py

```diff
@@ -0,0 +1,1072 @@
+	get_content: Callable[[DOMNode, Schema], Fragment] | None
+	preserve_whitespace: WSType
+
+	@classmethod
+	def from_json(cls, data: Dict[str, Any]) -> "ParseRule":
+		return ParseRule(
+			data.get('tag'),
+			data.get('namespace'),
+			data.get('style'),
+			data.get('priority'),
+			data.get('consuming'),
+			data.get('context'),
+			data.get('node'),
+			data.get('mark'),
+			data.get('clear_mark'),
+			data.get('ignore'),
+			data.get('close_parent'),
+			data.get('skip'),
+			data.get('attrs'),
+			data.get('getAttrs'),
+			data.get('contentElement'),
+			data.get('getContent'),
+			data.get('preserveWhitespace')
+		)
+
+
+class DOMParser:
+	_tags: List[ParseRule]
+	_styles: List[ParseRule]
+	_normalize_lists: bool
+
+	schema: Schema
+	rules: List[ParseRule]
+
+	def __init__(self, schema: Schema, rules: List[ParseRule]) -> None:
+		self.schema = schema
+		self.rules = rules
+		self._tags = [rule for rule in rules if rule.tag is not None]
+
+		self._styles = [rule for rule in rules if rule.style is not None]
+
+		self.normalize_lists = not any([
+			schema.nodes[r.node].content_match.match_type(schema.nodes[r.node])
+			for r in self._tags if r.node is not None and r.tag is not None and re.match(r'^(ul|ol)\b', r.tag) is not None
+		])
+
+	def parse(self, dom_: lxml.html.HtmlElement, options: ParseOptions | None = None) -> Node:
+		if options is None:
+			options = ParseOptions()
+
+		context = ParseContext(self, options, False)
+
+		for d in itertools.chain([dom_], dom_.iterdescendants()):
+			if d.text is not None and d.text.strip() and d.tag.lower() != 'lxmltext':
+				d.insert(0, lxml.html.fromstring(f'<lxmltext>{d.text}</lxmltext>'))
+				d.text = None
+
+			if d.tail is not None and d.tail.strip():
+				parent = d.getparent()
+				parent.insert(parent.index(d)+1, lxml.html.fromstring(f'<lxmltext>{d.tail}</lxmltext>'))
+				d.tail = None
+
+		context.add_all(dom_, options.from_, options.to_)
+
+		return cast(Node, context.finish())
+
+	def parse_slice(self, dom_: DOMNode, options: ParseOptions | None = None) -> Slice:
+		if options is None:
+			options = ParseOptions(preserve_whitespace=True)
+
+		context = ParseContext(self, options, True)
+
+		context.add_all(dom_, options.from_, options.to_)
+
+		return Slice.max_open(cast(Fragment, context.finish()))
+
+	def match_tag(self, dom_: DOMNode, context: "ParseContext", after: ParseRule | None = None) -> ParseRule | None:
+		try:
+			i = self._tags.index(after) + 1 if after is not None else 0
+		except ValueError:
+			i = 0
+
+		for rule in self._tags[i:]:
+			if (
+				rule.tag and matches(dom_, rule.tag)
+				and (rule.namespace is None or (dom_.prefix and dom_.nsmap[dom_.prefix] == rule.namespace))
+				and (not rule.context or context.matches_context(rule.context))
+			):
+				if rule.get_attrs is not None:
+					result = rule.get_attrs(dom_)
+					if result is False:
+						continue
+					rule.attrs = result
+
+				return rule
+
+			i += 1
```

Comment:
nit: Seems like we don't use `i` in the loop so should be no need to increment it